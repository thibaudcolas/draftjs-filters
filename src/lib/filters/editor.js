// @flow
import { EditorState } from "draft-js"
import type { DraftBlockType } from "draft-js/lib/DraftBlockType.js.flow"

import { ATOMIC, UNSTYLED } from "../constants"
import {
  preserveAtomicBlocks,
  resetAtomicBlocks,
  removeInvalidAtomicBlocks,
} from "./atomic"
import {
  limitBlockDepth,
  filterBlockTypes,
  removeInvalidDepthBlocks,
} from "./blocks"
import { filterInlineStyles } from "./styles"
import {
  filterEntityData,
  filterEntityRanges,
  shouldKeepEntityType,
  shouldRemoveImageEntity,
  shouldKeepEntityByAttribute,
} from "./entities"
import { replaceTextBySpaces } from "./text"

type FilterOptions = {
  // Whitelist of allowed block types. unstyled and atomic are always included.
  blocks: Array<DraftBlockType>,
  // Whitelist of allowed inline styles.
  styles: Array<string>,
  // Whitelist of allowed entities.
  entities: Array<{
    // Entity type, eg. "LINK"
    type: string,
    // Allowed attributes. Other attributes will be removed.
    attributes: Array<string>,
    // Refine which entities are kept by whitelisting acceptable values with regular expression patterns.
    whitelist: Object,
  }>,
  // Maximum amount of depth for lists (0 = no nesting).
  maxNesting: number,
  // Characters to replace with whitespace.
  whitespacedCharacters: Array<string>,
}

/**
 * Applies whitelist and blacklist operations to the editor content,
 * to enforce it's shaped according to the options.
 */
export const filterEditorState = (
  {
    blocks,
    styles,
    entities,
    maxNesting,
    whitespacedCharacters,
  }: FilterOptions,
  editorState: EditorState,
) => {
  const shouldKeepEntityRange = (content, entityKey, block) => {
    const entity = content.getEntity(entityKey)
    const entityData = entity.getData()
    const entityType = entity.getType()
    const blockType = block.getType()

    return (
      shouldKeepEntityType(entities, entityType) &&
      shouldKeepEntityByAttribute(entities, entityType, entityData) &&
      !shouldRemoveImageEntity(entityType, blockType)
    )
  }

  // Order matters. Some filters may need the information filtered out by others.
  const filters = [
    // 1. clean up blocks.
    removeInvalidDepthBlocks,
    limitBlockDepth.bind(null, maxNesting),
    // 2. reset styles and blocks.
    filterInlineStyles.bind(null, styles),
    // Add block types that are always enabled in Draft.js.
    filterBlockTypes.bind(null, blocks.concat([UNSTYLED, ATOMIC])),
    // 4. Process atomic blocks.
    preserveAtomicBlocks,
    resetAtomicBlocks,
    // 5. Remove entity ranges (and linked entities)
    filterEntityRanges.bind(null, shouldKeepEntityRange),
    // 6. Remove/filter entity-related matters.
    removeInvalidAtomicBlocks.bind(null, entities),
    filterEntityData.bind(null, entities),
    replaceTextBySpaces.bind(null, whitespacedCharacters),
  ]

  const content = editorState.getCurrentContent()
  const nextContent = filters.reduce((c, filter) => filter(c), content)

  return nextContent === content
    ? editorState
    : EditorState.set(editorState, {
        currentContent: nextContent,
      })
}
