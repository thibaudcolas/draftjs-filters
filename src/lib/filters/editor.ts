import {
  EditorState,
  ContentBlock,
  ContentState,
  CharacterMetadata,
} from "draft-js"
import { List } from "immutable"

import { ATOMIC, UNSTYLED } from "../constants"
import {
  preserveAtomicBlocks,
  resetAtomicBlocks,
  removeInvalidAtomicBlocks,
} from "./atomic"
import {
  BlockTextRule,
  limitBlockDepth,
  preserveBlockByText,
  filterBlockTypes,
  removeInvalidDepthBlocks,
} from "./blocks"
import { filterInlineStyles } from "./styles"
import {
  EntityRule,
  cloneEntities,
  filterEntityData,
  filterEntityRanges,
  shouldKeepEntityType,
  shouldRemoveImageEntity,
  shouldKeepEntityByAttribute,
} from "./entities"
import { replaceTextBySpaces } from "./text"
import { applyContentWithSelection } from "./selection"

interface FilterOptions {
  // List of allowed block types. unstyled and atomic are always included.
  blocks: readonly string[]
  // List of allowed inline styles.
  styles: readonly string[]
  // List of allowed entities.
  entities: readonly EntityRule[]
  // Maximum amount of depth for lists (0 = no nesting).
  maxNesting: number
  // Characters to replace with whitespace.
  whitespacedCharacters: readonly string[]
  // Optional: Rules used to switch block types based on block text.
  blockTextRules?: readonly BlockTextRule[]
}

const BLOCK_PREFIX_RULES = [
  {
    // https://regexper.com/#%5E(%C2%B7%20%7C%E2%80%A2%5Ct%7C%E2%80%A2%7C%F0%9F%93%B7%20%7C%5Ct%7C%20%5Ct)
    test: "^(Â· |â€¢\t|â€¢|ðŸ“· |\t| \t)",
    type: "unordered-list-item",
    depth: 0,
  },
  // https://regexper.com/#%5E(%E2%97%A6%7Co%20%7Co%5Ct)
  { test: "^(â—¦|o |o\t)", type: "unordered-list-item", depth: 1 },
  // https://regexper.com/#%5E(%C2%A7%20%7C%EF%82%A7%5Ct%7C%E2%97%BE)
  { test: "^(Â§ |ï‚§\t|â—¾)", type: "unordered-list-item", depth: 2 },
  {
    // https://regexper.com/#%5E1%7B0%2C1%7D%5Cd%5C.%5B%20%5Ct%5D
    test: "^1{0,1}\\d\\.[ \t]",
    type: "ordered-list-item",
    depth: 0,
  },
  {
    // Roman numerals from I to XX.
    // https://regexper.com/#%5Ex%7B0%2C1%7D(i%7Cii%7Ciii%7Civ%7Cv%7Cvi%7Cvii%7Cviii%7Cix%7Cx)%5C.%5B%20%5Ct%5D
    test: "^x{0,1}(i|ii|iii|iv|v|vi|vii|viii|ix|x)\\.[ \t]",
    type: "ordered-list-item",
    depth: 2,
  },
  {
    // There is a clash between this and the i., v., x. roman numerals.
    // Those tests are executed in order though, so the roman numerals take priority.
    // We do not want to match too many letters (say aa.), because those could be actual text.
    // https://regexper.com/#%5E%5Ba-z%5D%5C.%5B%20%5Ct%5D
    test: "^[a-z]\\.[ \t]",
    type: "ordered-list-item",
    depth: 1,
  },
] as const

/**
 * Applies filtering and preservation operations to the editor content,
 * to restrict it to supported patterns.
 * Will not alter the editor state if there are no changes to make.
 */
export const filterEditorState = (
  options: FilterOptions,
  editorState: EditorState,
) => {
  const {
    blocks,
    styles,
    entities,
    maxNesting,
    whitespacedCharacters,
    blockTextRules = BLOCK_PREFIX_RULES,
  } = options
  const shouldKeepEntityRange = (
    content: ContentState,
    entityKey: string,
    block: ContentBlock,
  ) => {
    const entity = content.getEntity(entityKey)
    const entityData = entity.getData()
    const entityType = entity.getType()
    const blockType = block.getType()

    return (
      shouldKeepEntityType(entities, entityType) &&
      shouldKeepEntityByAttribute(entities, entityType, entityData) &&
      !shouldRemoveImageEntity(entityType, blockType)
    )
  }

  // Order matters. Some filters may need the information filtered out by others.
  const filters = [
    // 1. clean up blocks.
    removeInvalidDepthBlocks,
    preserveBlockByText.bind(null, blockTextRules),
    limitBlockDepth.bind(null, maxNesting),
    // 2. reset styles and blocks.
    filterInlineStyles.bind(null, styles),
    // Add block types that are always enabled in Draft.js.
    filterBlockTypes.bind(null, blocks.concat([UNSTYLED, ATOMIC])),
    // 4. Process atomic blocks before processing entities.
    preserveAtomicBlocks,
    resetAtomicBlocks,
    // 5. Remove entity ranges (and linked entities)
    filterEntityRanges.bind(null, shouldKeepEntityRange),
    // 6. Remove/filter entity-related matters.
    removeInvalidAtomicBlocks.bind(null, entities),
    filterEntityData.bind(null, entities),
    // 7. Clone entities for which it is necessary.
    cloneEntities,
    // 8. Finally, do text operations.
    replaceTextBySpaces.bind(null, whitespacedCharacters),
  ]

  const content = editorState.getCurrentContent()
  const nextContent = filters.reduce((c, filter) => filter(c), content)

  return applyContentWithSelection(editorState, content, nextContent)
}

/**
 * Condenses an array of content blocks into a single block.
 * - Skipping the undo-redo stack.
 * - Adding a space between each block to match the behavior of vanilla HTML fields.
 * - Making sure the space gets the correct styles or entities applied.
 * - Placing the selection at the same end offset as the last selection.
 *
 * This exhibits two known issues:
 * - A link / other entity spread over multiple lines will be split into multiple entities of identical data.
 * - Upon redo-ing the change, the selection isnâ€™t correctly placed.
 */
export const condenseBlocks = (
  nextState: EditorState,
  prevState: EditorState,
) => {
  const content = nextState.getCurrentContent()
  const blockMap = content.getBlockMap()

  if (blockMap.size < 2) {
    return nextState
  }

  let text = ""
  let characterList: List<CharacterMetadata>

  // Calculate the condensed block text and character list,
  // making sure to insert a space between each block, with
  // the correct styles and entity.
  blockMap.forEach((block) => {
    // Atomic blocks should be ignored (stripped)
    if (block!.getType() !== "atomic") {
      text = `${text}${text ? " " : ""}${block!.getText()}`
      const blockList = block!.getCharacterList()
      characterList = characterList
        ? (characterList.concat(
            // Duplicate the CharacterMetadata of the blockâ€™s first character.
            blockList.slice(0, 1),
            blockList,
          ) as List<CharacterMetadata>)
        : blockList
    }
  })

  // Keep only the last block, with the combined text and character list.
  // Ideally we would keep the first block to have a stable key, but this
  // causes a runtime error when redo-ing the change.
  const changedBlocks = blockMap
    .slice(-1)
    .map((block) => block!.merge({ text, characterList }) as ContentBlock)

  const nextContent = content.merge({
    blockMap: changedBlocks,
  })

  // Calculate the new selectionâ€™s position, to be at the end of the pasted text.
  const prevSelection = prevState.getSelection()
  const prevOffset = Math.max(
    prevSelection.getFocusOffset(),
    prevSelection.getAnchorOffset(),
  )
  const prevContentLength = prevState
    .getCurrentContent()
    .getFirstBlock()
    .getLength()
  const newContentLength = text.length - prevContentLength

  const selection = nextState.getSelection()
  const nextSelection = selection.merge({
    anchorOffset: prevOffset + newContentLength,
    focusOffset: prevOffset + newContentLength,
  })
  const condensed = EditorState.set(nextState, {
    currentContent: nextContent,
  })

  return EditorState.acceptSelection(condensed, nextSelection)
}
