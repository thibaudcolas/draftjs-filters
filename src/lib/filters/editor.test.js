import { EditorState, convertFromRaw, convertToRaw } from "draft-js"

import { filterEditorState } from "./editor"

const preserveAtomicBlocksEntities = {
  "0": {
    type: "IMAGE",
    data: {
      src: "/0.png",
    },
  },
}

const preserveAtomicBlocks = [
  {
    key: "a",
    text: " ",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 0,
      },
    ],
  },
  {
    key: "b",
    text: "ðŸ“·",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 0,
      },
    ],
  },
  {
    key: "c",
    text: " ",
  },
  {
    key: "d",
    text: "ðŸ“· Star list",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 0,
      },
    ],
  },
]

const removeInvalidDepthBlocks = [
  {
    key: "e",
    text: "0",
    type: "ordered-list-item",
    depth: 0,
  },
  {
    key: "f",
    text: "1",
    depth: 1,
  },
  {
    key: "g",
    text: "2",
    type: "unordered-list-item",
    depth: 2,
  },
]

const resetBlockDepth = [
  {
    key: "h",
    text: "0",
    type: "unordered-list-item",
    depth: 0,
  },
  {
    key: "i",
    text: "1",
    type: "unordered-list-item",
    depth: 1,
  },
  {
    key: "j",
    text: "2",
    type: "unordered-list-item",
    depth: 2,
  },
]

const resetBlockType = [
  {
    key: "k",
    text: "P",
  },
  {
    key: "l",
    text: "UL",
    type: "unordered-list-item",
  },
  {
    key: "m",
    text: "H1",
    type: "header-one",
  },
]

const filterInlineStyle = [
  {
    key: "n",
    text: "tes",
    inlineStyleRanges: [
      {
        offset: 0,
        length: 1,
        style: "BOLD",
      },
      {
        offset: 1,
        length: 1,
        style: "ITALIC",
      },
      {
        offset: 2,
        length: 1,
        style: "BOLD",
      },
      {
        offset: 2,
        length: 1,
        style: "ITALIC",
      },
    ],
  },
]

const resetAtomicBlocksEntities = {
  "1": {
    type: "IMAGE",
    data: {
      src: "/1.png",
    },
  },
  "236": {
    type: "EMBED",
    data: {
      url: "http://www.youtube.com/watch?v=2",
    },
  },
}

const resetAtomicBlocks = [
  {
    key: "o",
    text: "ðŸ“·",
    type: "atomic",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 1,
      },
    ],
  },
  {
    key: "p",
    text: " ",
    type: "atomic",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 1,
      },
    ],
    inlineStyleRanges: [
      {
        offset: 0,
        length: 1,
        style: "BOLD",
      },
    ],
  },
  {
    key: "q",
    text: " ",
    type: "atomic",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 236,
      },
    ],
  },
  // TODO bug?
  {
    key: "r",
    text: " ",
    type: "atomic",
  },
]

const filterEntityRangesEntities = {
  "2": {
    type: "IMAGE",
    data: {
      src: "/example.png",
    },
  },
  "3": {
    type: "LINK",
    data: {
      url: "www.example.com",
    },
  },
  "328": {
    type: "TEST",
    data: {
      url: "doc.pdf",
    },
  },
  "512": {
    type: "EMBED",
    data: {
      url: "http://www.youtube.com/watch?v=y8Kyi0WNg40",
    },
  },
  "1024": {
    type: "IMAGE",
    data: {
      src: "http://example.com/example.png",
    },
  },
}

const filterEntityRanges = [
  // TODO Bug? Should drop the block if entity type is unsupported.
  {
    key: "s",
    text: " ",
    type: "atomic",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 512,
      },
    ],
  },
  {
    key: "t",
    text: " ",
    type: "atomic",
    entityRanges: [
      {
        offset: 0,
        length: 1,
        key: 2,
      },
    ],
  },
  {
    key: "u",
    text: "test",
    type: "unstyled",
    entityRanges: [
      {
        offset: 0,
        length: 2,
        key: 3,
      },
      {
        offset: 2,
        length: 2,
        key: 328,
      },
    ],
  },
  {
    key: "v",
    text: "shouldRemoveImageEntity ",
    type: "unstyled",
    entityRanges: [
      {
        offset: 23,
        length: 1,
        key: 2,
      },
    ],
  },
  // TODO Should remove atomic block coming from filtered-out entity.
  {
    key: "w",
    text: " ",
    type: "atomic",
    entityRanges: [
      {
        offset: 23,
        length: 1,
        key: 1024,
      },
    ],
  },
]

const filterEntityAttributesEntities = {
  "4": {
    type: "LINK",
    data: {
      href: "http://example.com",
      url: "http://example.com/",
    },
  },
}

const filterEntityAttributes = [
  {
    key: "x",
    text: "link",
    type: "unstyled",
    entityRanges: [
      {
        offset: 0,
        length: 4,
        key: 4,
      },
    ],
  },
]

const whitespaceCharacters = [
  {
    key: "y",
    text: "So\nft",
    type: "unstyled",
  },
  {
    key: "z",
    text: "Ta\tbs",
    type: "unstyled",
    inlineStyleRanges: [
      {
        offset: 0,
        length: 3,
        style: "BOLD",
      },
    ],
  },
]

describe("editor", () => {
  describe("#filterEditorState", () => {
    it("works", () => {
      const contentState = convertFromRaw({
        entityMap: Object.assign(
          {},
          preserveAtomicBlocksEntities,
          resetAtomicBlocksEntities,
          filterEntityRangesEntities,
          filterEntityAttributesEntities,
        ),
        blocks: [
          ...preserveAtomicBlocks,
          ...removeInvalidDepthBlocks,
          ...resetBlockDepth,
          ...resetBlockType,
          ...filterInlineStyle,
          ...resetAtomicBlocks,
          ...resetAtomicBlocks,
          ...filterEntityRanges,
          ...filterEntityAttributes,
          ...whitespaceCharacters,
        ],
      })
      const editorState = EditorState.createWithContent(contentState)
      expect(
        convertToRaw(
          filterEditorState({
            editorState,
            maxListNesting: 1,
            enableHorizontalRule: false,
            enableLineBreak: false,
            blockTypes: [
              "unstyled",
              "header-two",
              "header-three",
              "header-four",
              "unordered-list-item",
              "ordered-list-item",
            ],
            inlineStyles: ["BOLD"],
            entityTypes: [
              {
                type: "IMAGE",
                attributes: ["src"],
                whitelist: {
                  src: "^/",
                },
              },
              {
                type: "LINK",
                attributes: ["url"],
                whitelist: {},
              },
            ],
          }).getCurrentContent(),
        ),
      ).toMatchSnapshot()
    })

    it("enableHorizontalRule", () => {
      const editorState = EditorState.createEmpty()
      expect(
        filterEditorState({
          editorState,
          maxListNesting: 1,
          enableHorizontalRule: true,
          enableLineBreak: false,
          blockTypes: [],
          inlineStyles: [],
          entityTypes: [],
        }),
      ).toBeInstanceOf(EditorState)
    })

    it("enableLineBreak", () => {
      const editorState = EditorState.createEmpty()
      expect(
        filterEditorState({
          editorState,
          maxListNesting: 1,
          enableHorizontalRule: false,
          enableLineBreak: true,
          blockTypes: [],
          inlineStyles: [],
          entityTypes: [],
        }),
      ).toBeInstanceOf(EditorState)
    })

    it("no normalisation = no change", () => {
      const editorState = EditorState.createEmpty()
      expect(
        filterEditorState({
          editorState,
          maxListNesting: 1,
          enableHorizontalRule: false,
          enableLineBreak: true,
          blockTypes: [],
          inlineStyles: [],
          entityTypes: [],
        }),
      ).toBe(editorState)
    })
  })
})
