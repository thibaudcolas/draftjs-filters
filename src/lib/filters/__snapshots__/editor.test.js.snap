// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor #filterEditorState works 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 0,
          "length": 1,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "a",
      "text": " ",
      "type": "atomic",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 1,
          "length": 1,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "b",
      "text": " ",
      "type": "atomic",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "c",
      "text": " ",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d",
      "text": "Star list",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "e",
      "text": "0",
      "type": "ordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "g",
      "text": "2",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ab",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ac",
      "text": "Bullet 0",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ad",
      "text": "   ",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ae",
      "text": "Bullet 1",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "af",
      "text": "Bullet 2",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ag",
      "text": "Numbered 0",
      "type": "ordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ah",
      "text": "Numbered 2",
      "type": "ordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "ai",
      "text": "Numbered 1",
      "type": "ordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "h",
      "text": "0",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "i",
      "text": "1",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "j",
      "text": "2",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "k",
      "text": "P",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "l",
      "text": "UL",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "m",
      "text": "H1",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [
        Object {
          "length": 1,
          "offset": 0,
          "style": "BOLD",
        },
        Object {
          "length": 1,
          "offset": 2,
          "style": "BOLD",
        },
      ],
      "key": "n",
      "text": "tes",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 2,
          "length": 1,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "o",
      "text": " ",
      "type": "atomic",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 3,
          "length": 1,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "p",
      "text": " ",
      "type": "atomic",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 4,
          "length": 1,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "t",
      "text": " ",
      "type": "atomic",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 5,
          "length": 2,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "u",
      "text": "test",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "v",
      "text": "shouldRemoveImageEntity ",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 6,
          "length": 4,
          "offset": 0,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "x",
      "text": "link",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [
        Object {
          "key": 7,
          "length": 4,
          "offset": 0,
        },
        Object {
          "key": 8,
          "length": 4,
          "offset": 5,
        },
      ],
      "inlineStyleRanges": Array [],
      "key": "y",
      "text": "link link",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "z",
      "text": "So ft",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [
        Object {
          "length": 3,
          "offset": 0,
          "style": "BOLD",
        },
      ],
      "key": "aa",
      "text": "Ta bs",
      "type": "unstyled",
    },
  ],
  "entityMap": Object {
    "0": Object {
      "data": Object {
        "src": "/0.png",
      },
      "mutability": undefined,
      "type": "IMAGE",
    },
    "1": Object {
      "data": Object {
        "src": "/0.png",
      },
      "mutability": undefined,
      "type": "IMAGE",
    },
    "2": Object {
      "data": Object {
        "src": "/1.png",
      },
      "mutability": undefined,
      "type": "IMAGE",
    },
    "3": Object {
      "data": Object {
        "src": "/1.png",
      },
      "mutability": undefined,
      "type": "IMAGE",
    },
    "4": Object {
      "data": Object {
        "src": "/example.png",
      },
      "mutability": undefined,
      "type": "IMAGE",
    },
    "5": Object {
      "data": Object {
        "url": "www.example.com",
      },
      "mutability": undefined,
      "type": "LINK",
    },
    "6": Object {
      "data": Object {
        "url": "http://example.com/",
      },
      "mutability": undefined,
      "type": "LINK",
    },
    "7": Object {
      "data": Object {},
      "mutability": undefined,
      "type": "LINK",
    },
    "8": Object {
      "data": Object {},
      "mutability": undefined,
      "type": "LINK",
    },
  },
}
`;
